File: tests/test_ai_prompts.py
import json
from pathlib import Path
from scavenger.ai.prompts import build_prompt
from scavenger.models import Profile, Listing
from datetime import datetime, timezone

FIXTURES = Path(__file__).parent / "fixtures"


def make_profile() -> Profile:
    return Profile(
        id="sony", name="Sony A-mount Glass",
        keywords=["sony", ["a-mount", "alpha mount"]],
        negative_keywords=["broken", "fungus"],
        sources=["ebay"],
        price_min=50.0, price_max=800.0,
    )


def make_listing() -> Listing:
    now = datetime.now(timezone.utc)
    return Listing(
        id="abc", profile_id="sony", source_id="ebay",
        title="Sony 85mm f/1.4 A-mount",
        description="Great condition, no fungus. $249.99",
        price=249.99,
        url="https://ebay.com/1",
        first_seen=now, last_seen=now, relevance_score=80.0,
    )


def test_system_prompt_contains_profile_name():
    system, _ = build_prompt(make_profile(), make_listing())
    assert "Sony A-mount Glass" in system


def test_system_prompt_contains_keywords():
    system, _ = build_prompt(make_profile(), make_listing())
    assert "sony" in system.lower()
    assert "a-mount" in system.lower()


def test_system_prompt_contains_negative_keywords():
    system, _ = build_prompt(make_profile(), make_listing())
    assert "broken" in system.lower()
    assert "fungus" in system.lower()


def test_system_prompt_contains_json_schema():
    system, _ = build_prompt(make_profile(), make_listing())
    assert "relevant" in system
    assert "notable" in system
    assert "escalate" in system


def test_user_prompt_contains_listing_details():
    _, user = build_prompt(make_profile(), make_listing())
    assert "Sony 85mm f/1.4" in user
    assert "249.99" in user
    assert "no fungus" in user


def test_price_range_in_system_prompt():
    system, _ = build_prompt(make_profile(), make_listing())
    assert "50" in system
    assert "800" in system


def test_price_range_no_min_says_up_to():
    profile_no_min = Profile(
        id="sony2", name="Sony A-mount Glass",
        keywords=["sony"], negative_keywords=[],
        sources=["ebay"], price_max=800.0,
    )
    system, _ = build_prompt(profile_no_min, make_listing())
    assert "up to" in system
    assert "800" in system
    assert "$0" not in system


def test_fixture_is_valid_ai_evaluation():
    from scavenger.ai.models import AIEvaluation
    data = json.loads((FIXTURES / "ai_evaluation_sony.json").read_text())
    ev = AIEvaluation(**data)
    assert ev.relevant is True
    assert ev.notable is not None


# --- price range branch coverage ---

def _make_profile(**overrides):
    from scavenger.models import Profile
    defaults = dict(
        id="p1", name="Test", keywords=["test"],
        negative_keywords=[], sources=["ebay"],
    )
    defaults.update(overrides)
    return Profile(**defaults)


def _make_listing():
    from scavenger.models import Listing
    now = datetime.now(timezone.utc)
    return Listing(
        id="x", profile_id="p1", source_id="ebay", title="Test",
        url="https://test.com", first_seen=now, last_seen=now,
        relevance_score=50.0, price=100.0,
    )


def test_price_range_with_both_bounds():
    profile = _make_profile(price_min=50.0, price_max=500.0)
    system, _ = build_prompt(profile, _make_listing())
    assert "$50" in system and "$500" in system


def test_price_range_max_only():
    profile = _make_profile(price_max=500.0)
    system, _ = build_prompt(profile, _make_listing())
    assert "up to $500" in system


def test_price_range_min_only():
    profile = _make_profile(price_min=50.0)
    system, _ = build_prompt(profile, _make_listing())
    assert "$50 and above" in system


def test_price_range_neither():
    profile = _make_profile()
    system, _ = build_prompt(profile, _make_listing())
    assert "any price" in system


def test_batch_price_range_with_both_bounds():
    from scavenger.ai.prompts import build_batch_prompt
    profile = _make_profile(price_min=50.0, price_max=500.0)
    system, _ = build_batch_prompt(profile, [_make_listing()])
    assert "$50" in system and "$500" in system

File: rust/tests/test_ai_prompts.rs
use scavenger::ai::prompts::{build_batch_prompt, build_escalation_prompt, build_prompt};
use scavenger::models::{KeywordEntry, Listing, Profile};

fn test_profile() -> Profile {
    Profile {
        id: "p1".to_string(),
        name: "vintage synthesizers".to_string(),
        keywords: vec![
            KeywordEntry::Single("moog".to_string()),
            KeywordEntry::Group(vec!["minimoog".to_string(), "model d".to_string()]),
        ],
        negative_keywords: vec!["broken".to_string(), "parts only".to_string()],
        sources: vec!["ebay".to_string()],
        price_min: Some(500.0),
        price_max: Some(5000.0),
        poll_interval_sec: 3600,
        alert_priority: "normal".to_string(),
        enabled: true,
        tags: vec![],
        escalation_keywords: vec!["rare".to_string(), "mint".to_string()],
        location_radius_mi: None,
    }
}

fn test_listing() -> Listing {
    Listing {
        id: "abc123".to_string(),
        profile_id: "p1".to_string(),
        source_id: "ebay-12345".to_string(),
        title: "Vintage Minimoog Model D Synthesizer".to_string(),
        description: "Original 1972 Minimoog in great condition".to_string(),
        price: Some(3500.0),
        currency: "USD".to_string(),
        condition: Some("used".to_string()),
        url: "https://ebay.com/item/12345".to_string(),
        image_urls: vec![],
        location: None,
        first_seen: "2026-01-01T00:00:00Z".to_string(),
        last_seen: "2026-01-01T00:00:00Z".to_string(),
        relevance_score: 85.0,
        status: "new".to_string(),
        ai_evaluation: None,
    }
}

#[test]
fn build_prompt_includes_keywords() {
    let (system, _user) = build_prompt(&test_profile(), &test_listing());
    assert!(system.contains("moog"));
    assert!(system.contains("minimoog or model d"));
}

#[test]
fn build_prompt_includes_price_range() {
    let (system, _user) = build_prompt(&test_profile(), &test_listing());
    assert!(system.contains("$500"));
    assert!(system.contains("$5000"));
}

#[test]
fn build_prompt_includes_negative_keywords() {
    let (system, _user) = build_prompt(&test_profile(), &test_listing());
    assert!(system.contains("broken"));
    assert!(system.contains("parts only"));
}

#[test]
fn build_prompt_user_has_listing_details() {
    let (_system, user) = build_prompt(&test_profile(), &test_listing());
    assert!(user.contains("Vintage Minimoog Model D Synthesizer"));
    assert!(user.contains("$3500.00"));
    assert!(user.contains("Original 1972 Minimoog"));
}

#[test]
fn build_prompt_no_price() {
    let mut listing = test_listing();
    listing.price = None;
    let (_system, user) = build_prompt(&test_profile(), &listing);
    assert!(user.contains("price not listed"));
}

#[test]
fn build_prompt_no_description() {
    let mut listing = test_listing();
    listing.description = String::new();
    let (_system, user) = build_prompt(&test_profile(), &listing);
    assert!(user.contains("(no description)"));
}

#[test]
fn build_batch_prompt_includes_all_ids() {
    let listings = vec![
        test_listing(),
        {
            let mut l = test_listing();
            l.id = "def456".to_string();
            l.title = "Another synth".to_string();
            l
        },
    ];
    let (_system, user) = build_batch_prompt(&test_profile(), &listings);
    assert!(user.contains("[abc123]"));
    assert!(user.contains("[def456]"));
}

#[test]
fn build_batch_prompt_system_has_keywords() {
    let (system, _user) = build_batch_prompt(&test_profile(), &[test_listing()]);
    assert!(system.contains("moog"));
    assert!(system.contains("minimoog or model d"));
    assert!(system.contains("JSON array"));
}

#[test]
fn build_escalation_prompt_includes_triggered_keywords() {
    let triggered = vec!["rare".to_string(), "mint".to_string()];
    let (system, _user) = build_escalation_prompt(&test_profile(), &test_listing(), &triggered);
    assert!(system.contains("\"rare\""));
    assert!(system.contains("\"mint\""));
}

#[test]
fn build_escalation_prompt_includes_source() {
    let triggered = vec!["rare".to_string()];
    let (_system, user) = build_escalation_prompt(&test_profile(), &test_listing(), &triggered);
    assert!(user.contains("Source: ebay-12345"));
}

#[test]
fn build_escalation_prompt_deep_analysis() {
    let triggered = vec!["rare".to_string()];
    let (system, _user) = build_escalation_prompt(&test_profile(), &test_listing(), &triggered);
    assert!(system.contains("INSIDER KNOWLEDGE"));
    assert!(system.contains("MARKET CONTEXT"));
    assert!(system.contains("CREDIBILITY CHECK"));
}

#[test]
fn price_range_up_to_max_only() {
    let mut profile = test_profile();
    profile.price_min = None;
    profile.price_max = Some(1000.0);
    let (system, _user) = build_prompt(&profile, &test_listing());
    assert!(system.contains("up to $1000"));
}

#[test]
fn price_range_min_and_above() {
    let mut profile = test_profile();
    profile.price_min = Some(200.0);
    profile.price_max = None;
    let (system, _user) = build_prompt(&profile, &test_listing());
    assert!(system.contains("$200 and above"));
}

#[test]
fn price_range_any_price() {
    let mut profile = test_profile();
    profile.price_min = None;
    profile.price_max = None;
    let (system, _user) = build_prompt(&profile, &test_listing());
    assert!(system.contains("any price"));
}
